File: src/de/saschahlusiak/freebloks/view/effects/BoardStoneGlowEffect.java
package de.saschahlusiak.freebloks.view.effects;

import javax.microedition.khronos.opengles.GL10;

import de.saschahlusiak.freebloks.model.Stone;
import de.saschahlusiak.freebloks.view.BoardRenderer;
import de.saschahlusiak.freebloks.view.model.ViewModel;

public class BoardStoneGlowEffect extends AbsEffect implements Effect {
	int player, x, y;
	ViewModel model;
	Stone stone;
	
	private static final float BLEND_END = 0.55f;
	
	float color[] = new float[4];
	final float[] color1;
	final float[] color2;

	public BoardStoneGlowEffect(ViewModel model, int player, int x, int y, float distance) {
		this.model = model;
		this.player = player;
		this.x = x;
		this.y = y;
		this.time = -0.6f - distance * 0.025f;
		
		this.stone = new Stone();
		stone.init(0);
		
	    color1 = BoardRenderer.stone_color_a[player + 1];
	    color2 = BoardRenderer.stone_color_a[0];
	    
	    color[0] = color1[0];
	    color[1] = color1[1];
	    color[2] = color1[2];
	    color[3] = BoardRenderer.DEFAULT_ALPHA;
	}
	
	@Override
	public boolean isEffected(int x, int y) {
		return (x == this.x && y == this.y);
	}
	
	@Override
	public boolean isDone() {
		return time > BLEND_END;
	}
	
	@Override
	public boolean execute(float elapsed) {
	    float blend = (float)Math.sin(time / BLEND_END * Math.PI) * 0.9f;
	    if (time < 0.0f || time > BLEND_END)
	    	blend = 0.0f;
	    
	    
	    color[0] = color1[0] * (1.0f - blend) + color2[0] * blend;
	    color[1] = color1[1] * (1.0f - blend) + color2[1] * blend;
	    color[2] = color1[2] * (1.0f - blend) + color2[2] * blend;
	    color[3] = BoardRenderer.DEFAULT_ALPHA;
	    
		return super.execute(elapsed);
	}

	@Override
	public void render(GL10 gl, BoardRenderer renderer) {
		gl.glEnable(GL10.GL_BLEND);
		gl.glBlendFunc(GL10.GL_SRC_ALPHA, GL10.GL_ONE_MINUS_SRC_ALPHA);

		gl.glPushMatrix();
		
	    gl.glTranslatef(
	    		-BoardRenderer.stone_size * (float)(model.spiel.m_field_size_x - 1) + BoardRenderer.stone_size * 2.0f * (float)x,
	    		0,
	    		-BoardRenderer.stone_size * (float)(model.spiel.m_field_size_y - 1) + BoardRenderer.stone_size * 2.0f * (float)y);
		
		renderer.renderStone(gl, color);
		gl.glPopMatrix();

		gl.glDisable(GL10.GL_BLEND);
	}

	@Override
	public void renderShadow(GL10 gl, BoardRenderer renderer) {
		return;
	}
}

File: src/main/java/de/saschahlusiak/freebloks/view/effects/BoardStoneGlowEffect.kt
package de.saschahlusiak.freebloks.view.effects

import de.saschahlusiak.freebloks.Global
import de.saschahlusiak.freebloks.view.BoardRenderer
import de.saschahlusiak.freebloks.view.scene.Scene
import javax.microedition.khronos.opengles.GL10
import javax.microedition.khronos.opengles.GL11
import kotlin.math.sin

class BoardStoneGlowEffect(
    private var model: Scene,
    color: Int,
    private val x: Int,
    private val y: Int,
    distance: Float
) : AbsEffect() {
    private val timeMax = 0.55f

    private val color1 = Global.stone_color_a[color]
    private val color2 = Global.stone_color_a[0]
    private val color = FloatArray(4) { color1[it] * BoardRenderer.defaultStoneAlpha }

    init {
        time = -0.6f - distance * 0.025f
        this.color[3] = BoardRenderer.defaultStoneAlpha
    }

    override fun isEffected(x: Int, y: Int): Boolean {
        return x == this.x && y == this.y
    }

    override fun isDone(): Boolean {
        return time > timeMax
    }

    override fun execute(elapsed: Float): Boolean {
        var blend = sin(time / timeMax * Math.PI.toFloat()) * 0.9f

        if (time < 0.0f || time > timeMax) blend = 0.0f

        color[3] = BoardRenderer.defaultStoneAlpha
        color[0] = (color1[0] * (1.0f - blend) + color2[0] * blend) * color[3]
        color[1] = (color1[1] * (1.0f - blend) + color2[1] * blend) * color[3]
        color[2] = (color1[2] * (1.0f - blend) + color2[2] * blend) * color[3]

        return super.execute(elapsed)
    }

    override fun render(gl: GL11, renderer: BoardRenderer) {
        gl.glEnable(GL10.GL_BLEND)
        gl.glBlendFunc(GL10.GL_ONE, GL10.GL_ONE_MINUS_SRC_ALPHA)
        gl.glPushMatrix()
        gl.glTranslatef(
            -BoardRenderer.stoneSize * (model.board.width - 1).toFloat() + BoardRenderer.stoneSize * 2.0f * x.toFloat(),
            0f,
            -BoardRenderer.stoneSize * (model.board.height - 1).toFloat() + BoardRenderer.stoneSize * 2.0f * y.toFloat()
        )

        with (renderer) {
            gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_AMBIENT_AND_DIFFUSE, color, 0)
            gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SPECULAR, BoardRenderer.materialStoneSpecular, 0)
            gl.glMaterialfv(GL10.GL_FRONT_AND_BACK, GL10.GL_SHININESS, BoardRenderer.materialStoneShininess, 0)
            stone.bindBuffers(gl)
            stone.drawElements(gl, GL10.GL_TRIANGLES)
        }

        gl.glPopMatrix()
        gl.glDisable(GL10.GL_BLEND)
    }
}